clients/da-vinci-client/src/main/java/com/linkedin/davinci/store/rocksdb/RocksDBSstFileWriter.java

```diff
@@ -128,40 +132,71 @@ public void put(byte[] key, ByteBuffer valueBuffer) throws RocksDBException {
+   * If the number of SST files and the checkpoint do not match:
+   * 1. delete all the temporary SST files to be able to start ingestion from beginning
+   * 2. the files that are already ingested to the DB will be removed during re-ingestion
+   *    in {@link #deleteOldIngestion}
+   * 3. return false for the upstream to reset its state and restart ingestion.
+   */
+  boolean checkDatabaseIntegrity(Map<String, String> checkpointedInfo) {
     // Create temp SST file dir if it doesn't exist
     File tempSSTFileDir = new File(fullPathForTempSSTFileDir);
     if (!tempSSTFileDir.exists()) {
       tempSSTFileDir.mkdirs();
     }
+
     if (!checkpointedInfo.containsKey(lastCheckPointedSSTFileNum)) {
       LOGGER.info(
           "No checkpointed info for store: {}, partition id: {} so RocksDB will start building sst file from beginning",
           storeName,
           partitionId);
       lastFinishedSSTFileNo = -1;
       currentSSTFileNo = 0;
+      // Blindly remove all the temp sst files if found any. Will be recreated.
+      removeAllSSTFiles();
     } else {
       lastFinishedSSTFileNo = Integer.parseInt(checkpointedInfo.get(lastCheckPointedSSTFileNum));
       LOGGER.info(
           "Received last finished sst file no: {} for store: {}, partition id: {}",
           lastFinishedSSTFileNo,
           storeName,
           partitionId);
+
+      // This is not the first time this process is ingesting this partition,
+      // check the integrity before proceeding
       if (lastFinishedSSTFileNo < 0) {
         throw new VeniceException("Last finished sst file no: " + lastFinishedSSTFileNo + " shouldn't be negative");
       }
-      makeSureAllPreviousSSTFilesBeforeCheckpointingExist();
-      removeSSTFilesAfterCheckpointing();
-      currentSSTFileNo = lastFinishedSSTFileNo + 1;
+      if (doesAllPreviousSSTFilesBeforeCheckpointingExist()) {
+        // remove the unwanted sst files, as flow will continue from the checkpointed info
+        removeSSTFilesAfterCheckpointing(this.lastFinishedSSTFileNo);
+        currentSSTFileNo = lastFinishedSSTFileNo + 1;
+        LOGGER.info(
+            "Ingestion will continue from the last checkpoint for store: {} partition: {}",
+            storeName,
+            partitionId);
+      } else {
+        // remove all the temp sst files if found any as ingestion will be restarted from beginning
+        removeAllSSTFiles();
+        LOGGER.info("Ingestion will restart from the beginning for store: {} partition: {}", storeName, partitionId);
+        return false;
+      }
     }
+    return true;
+  }
+
+  public void open(Map<String, String> checkpointedInfo, Optional<Supplier<byte[]>> expectedChecksumSupplier) {
+    LOGGER.info(
+        "'beginBatchWrite' got invoked for RocksDB store: {}, partition: {} with checkpointed info: {} ",
+        storeName,
+        partitionId,
+        checkpointedInfo);
+    checkDatabaseIntegrity(checkpointedInfo);
```

Comment:
I suppose if `checkDatabaseIntegrity` returns `false`, this function should throw exception to error out.
In case the cleanup is done inside `checkDatabaseIntegrity`, but `open` is not aware, it could trigger some weird issue.

Reply:
Added a comment and throwing exception.